File: 1127-last-stone-weight/1127-last-stone-weight.py
import heapq

class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        maxHeap = []
        for stone in stones:
            heapq.heappush(maxHeap, -stone)
        
        while len(maxHeap) > 1:
            stone1 = (heapq.heappop(maxHeap)) * -1
            stone2 = (heapq.heappop(maxHeap)) * -1

            newStone = stone1 - stone2
            if newStone > 0:
                heapq.heappush(maxHeap, -newStone)
        
        return 0 if not maxHeap else -maxHeap[0]

        

File: 1127-last-stone-weight/1127-last-stone-weight.java
import java.util.PriorityQueue;
import java.util.Collections;

class Solution {
    public int lastStoneWeight(int[] stones) {

        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());

        for(int stone : stones) {
            maxHeap.offer(stone);
        }

        while(maxHeap.size() > 1) {
            int stone1 = maxHeap.poll();
            int stone2 = maxHeap.poll();

            int newStone = stone1 - stone2;
            
            if(newStone > 0) {
                maxHeap.offer(newStone);
            }
        }

        return maxHeap.size() == 0 ? 0 : maxHeap.peek();
        
    }
}